uasm: fix compilation on darwin; _7zz: cross-compilation fixes

thiagokokada opened the pull request:
###### Description of changes

Got `uasm` to build in `x86_64-darwin` by pulling this fix: https://github.com/Terraspace/UASM/pull/165

Also fixes `7zz` cross-compilation by setting `CC` and `CXX` flags correctly.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

thiagokokada committed:
uasm: fix compilation on darwin

thiagokokada commented on pkgs/development/compilers/uasm/default.nix:
```diff
@@ -14,9 +14,23 @@ stdenv.mkDerivation rec {
+    else
+      "gccLinux64.mak";
+
+  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
```
Maybe this will fix cross-compilation too, but I didn't test this actually.

thiagokokada replied:
Seems it works:

```
~/Projects/nixpkgs fix-uasm-macos
❯ nix build .\#pkgsCross.aarch64-multiplatform.uasm

~/Projects/nixpkgs fix-uasm-macos 40s
❯ ./result/bin/uasm
UASM v2.55, Jan  1 1980, Masm-compatible assembler.
Portions Copyright (c) 1992-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.

usage: UASM [ options ] filelist [@env_var]
Run "Uasm -?" or "Uasm -h" for more info
```

(Not sure if it didn't work before though).

thiagokokada replied:
```
~/Projects/nixpkgs master
❯ nix build .\#pkgsCross.aarch64-multiplatform.uasm
error: builder for '/nix/store/jg130hrb4dnai8ndfwa49xknxi48v6cs-uasm-aarch64-unknown-linux-gnu-2.55.drv' failed with exit code 2;
       last 10 log lines:
       > make: *** Waiting for unfinished jobs....
       > make: *** [gccLinux64.mak:35: GccUnixR/apiemu.o] Error 127
       > /nix/store/fcd0m68c331j7nkdxvnnpb8ggwsaiqac-bash-5.1-p16/bin/bash: line 1: gcc: command not found
       > make: *** [gccLinux64.mak:35: GccUnixR/orgfixup.o] Error 127
       > /nix/store/fcd0m68c331j7nkdxvnnpb8ggwsaiqac-bash-5.1-p16/bin/bash: line 1: gcc: command not found
       > make: *** [gccLinux64.mak:35: GccUnixR/assemble.o] Error 127
       > /nix/store/fcd0m68c331j7nkdxvnnpb8ggwsaiqac-bash-5.1-p16/bin/bash: line 1: gcc: command not found
       > make: *** [gccLinux64.mak:35: GccUnixR/assume.o] Error 127
       > /nix/store/fcd0m68c331j7nkdxvnnpb8ggwsaiqac-bash-5.1-p16/bin/bash: line 1: gcc: command not found
       > make: *** [gccLinux64.mak:35: GccUnixR/atofloat.o] Error 127
       For full logs, run 'nix log /nix/store/jg130hrb4dnai8ndfwa49xknxi48v6cs-uasm-aarch64-unknown-linux-gnu-2.55.drv'.
```

So yeah, it didn't before. Nice bonus.

thiagokokada commented:
@ofborg build _7zz

thiagokokada commented:
Result of `nixpkgs-review pr 169232` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_7zz</li>
    <li>uasm</li>
  </ul>
</details>

thiagokokada committed:
_7zz: cross-compilation fixes

siraben commented:
Result of `nixpkgs-review pr 169232` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>uasm</li>
  </ul>
</details>

thiagokokada committed:
uasm: add testVersion

thiagokokada committed:
_7zz: useUasm only in x86 platforms

`uasm` is x86 only.

It seems that in `aarch64-linux` at least, the optimizing build is done
without using any third-party tool (maybe using GCC's own assembly?).

06kellyjac commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -69,17 +77,13 @@ stdenv.mkDerivation rec {
-  doInstallCheck = true;
-
-  installCheckPhase = ''
-    runHook preInstallCheck
-
-    $out/bin/7zz --help | grep ${version}
-
-    runHook postInstallCheck
-  '';
-
-  passthru.updateScript = ./update.sh;
+  passthru = {
+    updateScript = ./update.sh;
+    tests.version = testVersion {
+      package = _7zz;
+      command = "7zz --help";
+    };
+  };
```
> installCheckPhase is similar to checkPhase, in what both means that if this test fail I don't even want this package to be build

> But if the idea is test only the current package, installCheckPhase generally makes more sense to me (of course, there could be exceptions).

https://github.com/NixOS/rfcs/pull/119#discussion_r783597367

I'd like to keep the `installCheckPhase` as-is for now until RFC0119 is ironed out.

thiagokokada replied:
Well, I followed most discussions in RFC0119 and it seems that the general perception is that `passthru.tests` and `installCheckPhase` should be considered the same: the package shouldn't be merged if it fails. @ofborg will automatically run `passthru.tests`, and nobody should merge packages that don't pass the tests without a good reason.

As an argument for `testVersion`: it does more than the previous test and it is more standard, so I am in favor of keeping it.

thiagokokada replied:
For things `testVersion` does better than the previous test:

- It checks if the return code of `command` is 0, otherwise the test fails
- It uses the `grep -Fw` instead of just `grep`. This avoids matching something like `Random string with the version21.07mixed inside just on coincidence`.

thiagokokada replied:
And for development, I found it sufficient to run `nix-build -A _7zz.tests` instead of `nix-build -A _7zz`. This will run the tests, but since the tests depend on the package itself, they will also retrigger the package rebuild.

06kellyjac commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -2,23 +2,27 @@
+    i686-linux = "_x86";
+    x86_64-linux = "_x64";
+  }.${system} or
+    (builtins.trace "`platformSuffix` not available for `${system}.` Making a generic `7zz` build." "");
```
Is this clarification in the wording because the generic build can have ASM optimizations too?
The profile specific make files do have `USE_ASM=1`. If we want the makefiles to be detached from whether we use ASM or not we probably want to control `USE_ASM` based on `useUasm`

thiagokokada replied:
No, the generic build doesn't have ASM optimizations. All the other ones (the architecture specific ones) seems to have, however it is not a simple case of setting `USE_ASM` because some architectures use third party tools to generate the assembly code (e.g. `x86_64` uses `uasm`), while others doesn't (e.g.: `aarch64-linux`).

06kellyjac replied:
ok cool so `USE_ASM` is general and `useUasm` is for using `usam`

:+1:

06kellyjac commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -2,23 +2,27 @@
   # RAR code is under non-free unRAR license
   # see the meta.license section below for more details
 , enableUnfree ? false
+
+  # For tests
+, _7zz
+, testVersion
```
See passthru comment

06kellyjac commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -50,6 +54,10 @@ stdenv.mkDerivation rec {
 
   makeFlags =
+    [
+      "CC=${stdenv.cc.targetPrefix}cc"
```
There's a `CROSS_COMPILE` makeflag/var

I guess we could do `"CROSS_COMPILE=${stdenv.cc.targetPrefix}"`? but then we might also want to turn it on only for cross compiling?

thiagokokada replied:
I don't think we need. Probably `CROSS_COMPILE` is for a specific need for cross compilation, but just setting `CC` and `CXX` seems sufficient to get a working cross compile build for me.

And why set this regardless variables regardless? This is considered best practice in `nixpkgs`. Also, there is no issue of this causing our build to use a compiler we don't expect (like in other systems), thanks to the Nix hermetic build system.

thiagokokada replied:
BTW, if `CROSS_COMPILE` adds the prefix to the compiler, setting it alone wouldn't work. We would also need to set both `CC=cc` and `CXX=c++` since we can't assume that the C/C++ compiler will always be GCC. That is, unless the package only builds with GCC, them this is probably a better solution, but AFAIK `7zz` builds happily with Clang (at least the generic build).

06kellyjac replied:
Looks like it's just for prefix

```
λ rg CROSS_COMPILE ../../..
../../../7zip/var_clang.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_clang_x64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_clang_arm64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_clang_x86.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_gcc_x64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)gcc
10:CXX=$(CROSS_COMPILE)g++

../../../7zip/var_mac_arm64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_gcc_x86.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)gcc
10:CXX=$(CROSS_COMPILE)g++

../../../7zip/var_mac_x64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)clang
10:CXX=$(CROSS_COMPILE)clang++

../../../7zip/var_gcc.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)gcc
10:CXX=$(CROSS_COMPILE)g++

../../../7zip/var_gcc_arm64.mak
6:CROSS_COMPILE=
9:CC=$(CROSS_COMPILE)gcc
10:CXX=$(CROSS_COMPILE)g++
```

---

Is there a reason `"CC=${stdenv.cc.targetPrefix}cc"` etc isn't just default stdenv behavior in nixpkgs if it's used so often?

thiagokokada replied:
> Is there a reason "CC=${stdenv.cc.targetPrefix}cc" etc isn't just default stdenv behavior in nixpkgs if it's used so often?

No idea.

I would assume that this was something that they eventually decided it was a good idea, but turning by default would break tons of things. But I am only assuming this, I have no idea if this is true or not.

thiagokokada replied:
> I guess we could do "CROSS_COMPILE=${stdenv.cc.targetPrefix}"? but then we might also want to turn it on only for cross compiling?

If you do want to use it instead for some reason (since it has this flag maybe upstream may use it for something else in the future), I think we can do this change. No need to set it conditionally though, if there is no cross compilation happening `stdenv.cc.targetPrefix` will be set to empty string.

06kellyjac replied:
ok cool. well it looks like `CROSS_COMPILE` would be the 7zz maintainer's "preferred" way of adding this but it's much more nixpkgs standard to have it this way

Although as a note there are a handful of packages using `CROSS_COMPILE`

```
λ rg '"CC=\$\{stdenv\.cc\.targetPrefix\}cc"' | wc -l
111
λ rg '"CROSS_COMPILE=\$\{stdenv\.cc\.targetPrefix\}"' | wc -l
15
```

06kellyjac commented:
I'd appreciate a say before merging. Especially if the PR has been up for less than 12h

thiagokokada commented:
> I'd appreciate a say before merging. Especially if the PR has been up for less than 12h

Can you please add yourself as a maintainer them?

06kellyjac commented:
I am?

https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/7zz/default.nix#L98

![image](https://user-images.githubusercontent.com/9866621/163958697-c30b33e7-5302-4c1e-ad3a-85e9e2e6776c.png)

thiagokokada commented:
> I am?
> 
> https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/7zz/default.nix#L98
> 
> ![image](https://user-images.githubusercontent.com/9866621/163958697-c30b33e7-5302-4c1e-ad3a-85e9e2e6776c.png)

Sorry. It is really difficult to know people that use completely different maintainer identifier from their GitHub handles.

thiagokokada commented:
@06kellyjac Anyway, back to the discussion. I answered most of your objections and I still think that my changes should stay, but if you still don't agree with me I can undo some of them.

SuperSandro2000 commented on pkgs/development/compilers/uasm/default.nix:
```diff
@@ -14,9 +20,23 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-CIbHPKJa60SyJeFgF1Tux7RfJZBChhUVXR7HGa+gCtQ=";
   };
 
+  patches = lib.optionals stdenv.isDarwin [
```
This patch is not touching linux files so making it not conditional doesn't break anything. Making patches conditional increases the chance that with the next update the patch is forgotten when it no longer applies.

thiagokokada replied:
I expect this patch to be removed on the next release though.

SuperSandro2000 replied:
Thats great but if someone else updates uasm and is not on Mac the chance to miss this are pretty realistic especially when using nix-update.

thiagokokada replied:
Sure. If this happens we fix it when it happens.

I don't think we should bother opening another PR just to fix a possible issue.

SuperSandro2000 replied:
> Sure. If this happens we fix it when it happens.

Sure but most likely it will only be discovered X days/months later which puts unnecessary extra maintenance work onto us.



> Anyway, I will try to avoid doing this in the future (even if I don't like applying random patches unless strictly necessary).

That would be great!

thiagokokada replied:
> Sure but most likely it will only be discovered X days/months later which puts unnecessary extra maintenance work onto us.

I still think this wouldn't actually happen since we still have `{aarch64,x86_64}-darwin` running in CI. The only way I can see this happening is someone bumping this and ignoring the CI checks (that I know it happens sometimes, but we shouldn't).

thiagokokada replied:
Also, let's say this patch does affect other platforms but is only necessary for Darwin? Should we still apply it globally, having the risk of introducing bugs on all platforms because of one platform (that is known to be problematic in `nixpkgs`)?

If not, them we come again to the same issue that if we don't want to break Darwin we need to open the PR and wait for the CI checks at least. This is why I much prefer to apply patches only when strictly necessary, because I think supporting Darwin is good to have, but not essential.

thiagokokada replied:
Anyway, I think this discussion extended too much. Is there some other place that we could discuss those things?

SuperSandro2000 replied:
> ignoring the CI checks (that I know it happens sometimes, but we shouldn't).

If I am in a hurry I usually tend to more easily ignore darwin failures. 

> Also, let's say this patch does affect other platforms but is only necessary for Darwin? Should we still apply it globally, having the risk of introducing bugs on all platforms because of one platform (that is known to be problematic in `nixpkgs`)?

I think that depends on the case. Often patches are only for some platform specific code.

---


You can chat me up on matrix if you like.

thiagokokada replied:
> If I am in a hurry I usually tend to more easily ignore darwin failures.

If you care about Darwin, I think you shouldn't. Any change is a possible breakage, so the only way to ensure that nothing is breaking is at least leaving the CI to run until it finishes.

If you don't care, it is fine. But I don't think that having non conditional patches will help this that much, specially in one package that probably nobody cares. I (as the maintainer) particularly prefer that this package was updated even if the Darwin build was broken (for example, via @r-ryantm bot), because I don't have another way to be notified about packages updates otherwise.

SuperSandro2000 replied:
> Was going to send you a Matrix message but I don't know your handle.

``@sandro:supersandro.de``

SuperSandro2000 commented on pkgs/tools/archivers/7zz/default.nix:
```diff
@@ -2,23 +2,27 @@
 , lib
 , fetchurl
 
+  # Only used for x86/x86_64
```
It is very likely that the comment will be forgotten to be updated when something update the condition changes.

thiagokokada replied:
It is very unlikely that this will ever change though. `uasm` only builds x86 and x86_64 assembly, and upstream said they're not interested of adding any new backends.

Also, even if `uasm` added support for, say, `aarch64`, I don't think `7zz` would switch to it since they already wrote the code in `.S` format.
